Team 5 Enemy - Self Destruction Drones, Changes to Bomber, Bug Fixes with Activation, Drone Death

keerthanathummala opened the pull request:
# Description

### Self Destruction Drone
This PR introduces a new Self-Destruction Drone variant.  The attached `selfDestructComponent` allows the drone to self-destruct when:
- They collide with player (within a defined radius) or,
- It moves beyond the maximum distance from the player

Upon triggering the drone : 
- Deals damage to the player
- Plays an explosion animation
- Plays and fades out the explosion sound
- Neutralises physics and disables the entity after the effect (event driven on explosion animation end)

Note: The collider outline appears in the debug renderer but physics is disabled so no unexpected interactions. On level reset, the normal disposal path runs and the self-destruct drone respawns as expected.

### Bomber Drone with Spotlight Detection
This PR also updates the Bomber's animations and detection methods from position detection to cone light detection.
The updated bomber animations allow players to more easily distinguish between regular drone enemies and bomber drones. Furthermore, the optimised animations provide smoother movement and bomb-dropping actions for the bomber drones.

The enhanced cone-shaped light detection system gives players a clearer understanding of the bomber drone's target detection range. Once the target gets into the cone light detection area, the bomber will attack immediately. Each Bomber drone has a unique cone light number so that it can make every drone independent of the others. 

For more information, please see: https://github.com/UQcsse3200/2025-studio-2/issues/137

**Expected Behaviour for Bomber Variants**
* **Patrol Bomber (cone-detect):** Drops only when player is inside the cone, bomber is at optimal height and roughly aligned horizontally. Respects cooldown, bombs are tracked and cleaned up on reset. Default behaviour is patrol, but bomber will switch to engaging the player when it is detected by the attached cone light.
* **Auto-Bomber (route-only):** Patrols a provided route and **auto-drops** on a fixed cadence (e.g. 2 seconds) regardless of player presence. Also uses the `BombTrackerComponent` to ensure proper cleanup on reset.

### Bug Fixes to Activation Component
This PR also updates the ActivationComponent to address a bug where enemies were no longer re-activated by security cameras after a manual level reset. The fix ensures that activation events are correctly re-registered on resets (manual and after player death) and includes additional unit tests in `ActivationComponentTest`.

### Drone Death upon Trap/Death Zone (#129 )
Introduces implementation of drone death when a drone collides with a puzzle trap or a death zone and then explodes. This feature allows players to destroy drones through environmental hazards which adds to strategic gameplay.

## Upgraded Gradle
The upgraded Gradle version 8.7 can directly package using the GDX Texture Packer functionality without affecting the implementation or usage of other features.

## Type of change
- [x] new Feature 

# How Has This Been Tested?
- [x] Spawned a drone near the player and confirmed explosion occurs on collision within 1.3f radius
- [x] Moved the player more than 30f away from the drone confirmedthe rmed automatic explosion trigger
- [x] Verified players'the  health decreases by 2 when explosion is triggered
- [x] Confirmed explosion animation plays and stops after 0.5sec
- [x] Explosion sound plays and fades out over 0.5 sec
- [x] Entity component(`Physics`, `AnimationRenderComponent`, `SelfDestructionComponent`) is. The removed and destroyed event is triggered

# Checklist:
- [x] Code follows maximum Self-Destruction of armed sp triggering  the guidelines and understands players, confirmed the armedsp triggering the maximum Self-Destruction of the style guidlines of this project
- [x] Self-review of code has been performed
- [x] Code has been commented, particularly in hard to understand areas
- [x] Changes generate no new warnings
- [x] Tests have been added tp prove the feature works
- [x] New and existing unit tests pass locally with these changes
- [x] Any dependt changes have been merged and published in downstream modules

kanavgulati19 committed:
Implemented component of self-destruct drone,,SelfDestructComponent

kanavgulati19 committed:
Added self destruction drone in EnemyFactory

kanavgulati19 committed:
Updated ForestGameArea with spawnSelfDestructDrone method

kanavgulati19 committed:
called the spawnSelfDestructDrone function

kanavgulati19 committed:
Updated self destruction drone speed

keerthanathummala committed:
solving merge conflict

keerthanathummala committed:
Merge branch 'self-destruction-drone' of https://github.com/UQcsse3200/2025-studio-2 into self-destruction-drone
s the commit.

keerthanathummala committed:
added seld Destruction Drone image

keerthanathummala committed:
resolved conflicts after pulling content from team_5_enemy branch

keerthanathummala committed:
created SelfDrestructionDrone atlas file

keerthanathummala committed:
loaded SelfDestructDrone.atlas file

keerthanathummala committed:
spawning self destruction drone

keerthanathummala committed:
few errors occured while merging file ,resolved conflicts by added the missing arguments

kanavgulati19 committed:
Conflicts resolved in ForestGameArea

keerthanathummala committed:
trying to solve conflicts

keerthanathummala committed:
changes made to SelfDestructioncomponent code to animate flying and self_destruct

keerthanathummala committed:
added SelfDestructDrone tests

kanavgulati19 committed:
Saved changes

keerthanathummala committed:
resolved merge conflict

keerthanathummala committed:
resolved merge conflict

kanavgulati19 committed:
Resolved Conflicts

kanavgulati19 committed:
Resolved Conflicts

kanavgulati19 committed:
Restored EnemyFactory.java and ForestGameArea.java from commit 643e1d9

kanavgulati19 committed:
combined local changes before rebase

keerthanathummala committed:
after reseting the commits, retying missing components

kanavgulati19 committed:
Updated enemyfactory

keerthanathummala committed:
fixed errors

keerthanathummala committed:
resilved confilcts

kanavgulati19 committed:
removed selfdestruct.atlas and only kept drone.atlas in explosionfactory

kanavgulati19 committed:
Spawned drone in sprint1game area and activated chasetask

kanavgulati19 committed:
Merge remote-tracking branch 'refs/remotes/origin/self-destruction-drone' into self-drone-destruction

kanavgulati19 committed:
Updated drone to explode while near player

kanavgulati19 committed:
Updated drone to self-destruct

keerthanathummala committed:
tried to reslove bugs which were introduced while creation of new atlas file

kanavgulati19 committed:
Merged self-destruction-drone with team_5_enemy

kanavgulati19 committed:
Fixed minor errors and pushed the code to self-destruction-drone branch

ROSE16111 committed:
upgrade gradlew wrapper to 8.7 AND add drone teleport animation

charlottemaherr committed:
Add patrol drone and security light in LevelOne for testing

charlottemaherr committed:
Add teleport wait subtask in Cooldown so teleport animation plays right before respawn

kanavgulati19 committed:
Merge remote-tracking branch 'origin/team_5_enemy' into self-destruction-drone

kanavgulati19 committed:
Spawned Self-destruct-drone in LevelOneGameArea

Russel1lTan committed:
Implemented the bomber drone with fully functional and light cone detect

keerthanathummala committed:
added test cases for selfDestructDrone

keerthanathummala committed:
changed position of SelfDestructDrone

keerthanathummala committed:
resolved merge conflicts

keerthanathummala committed:
changes made

keerthanathummala committed:
errors resolve

kanavgulati19 committed:
Resolved Conflicts

Russel1lTan committed:
Implemented the bomber drone with fully functional and light cone detect, and fixed the EnemyFactoryTest

kanavgulati19 committed:
Changed max speed

kanavgulati19 committed:
Developed self destructing drone which deals damage to the player  by exploding once at a particular distance closest to the player

kanavgulati19 committed:
Resolved conflicts

policarmahalia committed:
increase knockback force to make effect visible

keerthanathummala committed:
Merge branch 'team_5_enemy' of https://github.com/UQcsse3200/2025-studio-2 into self-destruction-drone

keerthanathummala committed:
updated Test Cases of self destruct drone

keerthanathummala committed:
activated bomber drone

keerthanathummala committed:
changed position of SelfDestructionDrone

ROSE16111 committed:
add Bomber teleport animation in atlas file

ROSE16111 committed:
add prameter in cooldown task to control teleport animation type

ROSE16111 committed:
add self Explosion animation

policarmahalia committed:
Implemented drone death in collision to death zones and traps

policarmahalia committed:
Added tests for DeathOnTrapComponent

policarmahalia committed:
Merge branch 'team_5_enemy' of https://github.com/UQcsse3200/2025-studio-2 into team_5_enemy

ROSE16111 committed:
Merge branch 'team_5_enemy' into self-destruction-drone

keerthanathummala committed:
made changes to selfDestruction drone so that after explosion it doesn't affect rest of the drones

ROSE16111 committed:
Merge pull request #131 from UQcsse3200/self-destruction-drone

Self destruction drone

Russel1lTan committed:
Added more bomber animations and created the BombPatrolTask based on PatrolTask. Modified EnemyFactoryTest

keerthanathummala committed:
resolved bug related to explosion of self-destruction drone and pulled latest version of the game from main

Russel1lTan committed:
Merged main into t5 branch and fixed EnemyFactoryTest error

keerthanathummala committed:
fixed an issue(initially after the explosion the remains of explosion remains but not anymore)

keerthanathummala committed:
made changes so that after explosion health is droped by 2 hearts

keerthanathummala committed:
changes made to explode function after manual testing

keerthanathummala committed:
added background sound to explosion

kanavgulati19 committed:
Solved conflicts

kanavgulati19 committed:
Updated createSelfDestructionDrone with light component in EnmeyFactory

kanavgulati19 committed:
Added light chase

kanavgulati19 committed:
Updated selfdestructcomponent to dispose the light after explosion

keerthanathummala committed:
add Self destruction drone to all the levels

keerthanathummala committed:
pulled latest changes from selff-destruction-drone branch

keerthanathummala committed:
made changes , so now when player is 11f away from selfDestruction Drone(SDD) the drone teleports near the play and then when reaches player it explodes

charlottemaherr committed:
Updated activation component to handle re-linking on level reset

charlottemaherr committed:
Updated activation component to handle re-linking on level reset

charlottemaherr committed:
Add security camera activation for self destruct drones. Replace teleport with chase + cooldown

charlottemaherr committed:
Link self destruct drone to security camera in levelOneGameArea

kanavgulati19 committed:
updated teleport task and changed speed of drone

kanavgulati19 committed:
Added cone light below drone for more dynamic gameplay

keerthanathummala committed:
made changes so the drone only teleports once it started chasing

keerthanathummala committed:
The SDD explode when Player is with in the collision radius instead if 1.1f

keerthanathummala committed:
changed collision radius

kanavgulati19 committed:
Updated self-destruct component

keerthanathummala committed:
commenting out teleport function of drone

keerthanathummala committed:
increasing the range of chase between player and drone

keerthanathummala committed:
updated SDD

kanavgulati19 committed:
Updated SDD so that it will immediately start chasing the player on spawn in boss level once it is created

kanavgulati19 committed:
Reverted some changed to SDD

kanavgulati19 committed:
Wrote tests for SelfDestructComponent

kanavgulati19 committed:
Merge remote-tracking branch 'origin/team_5_enemy' into self-destruction-drone

kanavgulati19 committed:
Removed the conelight under the drone

kanavgulati19 committed:
Updated SDD tests

Russel1lTan committed:
Solved bomb dispose bug and created an automatic bomber drone for boss generate drone

Russel1lTan committed:
Added the boss animations for Chase, Touch Kill

Shane-Siri reviewed (changes requested):
Im in your branch (team_5_enemy) and I did a clean building and ran the code. Its seems that the the level two area doesnt have does assets of altas' and doesnt load? Is this something you are experiencing too?

Russel1lTan committed:
Fixed Level two loading error and javadoc.io error.

jacobem1836 reviewed (approved):
Looks good! Everything working as it should - only thing is if you hang out above the bomber drones, they will attempt to drop a bomb 'upwards' and damage themselves. Approving as its not a major functionality issue and could be valid gameplay maybe?

Edit: see below

jacobem1836 reviewed (changes requested):
Crashes when first self destruct drone in level 2 explodes

alicewalker1 reviewed (changes requested):
Everything looks super cool but I did find one kind of large bug. When the first self destruct drone on level 2 explodes it freezes the entire game and it stops responding. This is the best image I could get of it happening:
<img width="1922" height="1112" alt="image" src="https://github.com/user-attachments/assets/b726537d-b5ab-457d-ab97-abbbf64f9877" />
Once this is fixed I'm happy to approve

tristynhunt reviewed (changes requested):
Bomber drone projectiles spam colliders and they don't despawn even after a level restart. This causes a large build up of colliders. There is also a secondary smaller collider which the player can interact with to the right of the main one. It's weirdly bouncy too and leads to some very unexpected interactions that shouldn't be happening.
<img width="1234" height="746" alt="image" src="https://github.com/user-attachments/assets/464f7a08-8ab7-44c5-b026-e915b81c7a31" />

charlottemaherr committed:
Fix position of bomb projectile's collider component

keerthanathummala committed:
the drone is disposed

charlottemaherr committed:
Fix: Bombs dispose properly after explosion

policarmahalia committed:
Fixed issue causing game crash when drones exploded using SelfDestructComponent logic.

policarmahalia committed:
Refactor DeathOnTrapComponent to reuse explode logic and remove components instead of disposing entity. Added javadocs.

policarmahalia committed:
Bug fix: update SelfDestructComponent to fully remove entity after explosion

charlottemaherr committed:
Add BombTrackerComponent to fix bomb disposal and reset behaviour

Defers bomb disposal to next tick and ensures all unexploded bombs are disposed properly on level reset

charlottemaherr committed:
Fix: Forgot to stage BombTrackerComponent

charlottemaherr committed:
Fix issue with ghost collider and clean up after self destruct drone explosion

charlottemaherr committed:
Merge remote-tracking branch 'origin/main' into team_5_enemy

charlottemaherr commented:
Thanks everyone for the reviews! After a huge night of bug fixing we've resolved:

1. Level 2 freeze when self destruct drone explodes
Switched to event-driven cleanup after explosion animation ends. Instead of tearing down drone components mid-update, we've neutralised the physics to prevent unexpected collisions in `selfDestructComponent`. Just a note - the debug renderer still shows the collider outline for now but since physics is disabled there are no unexpected interactions. On level reset, normal disposal path runs and self destruct respawns as expected.

2. Pre-loaded  `sounds/explosion.mp3' and drone atlas in Level 2 to fix errors with asset loading.

3. Bomber projectile colliders persisting after bomb explosion and level resets
This was a bigger change - we introduced a new component for bomber drones that tracks active bombs and ensures they are correctly disposed on level reset events. Now when we manually reset a level any unexploded bombs are disposed and there are no extra colliders. More details on what was changed here: #190 
**Note on reset warning:** When we trigger level reset and bombs are disposed, we get the severe warning about entities leaked. This is expected because bombs are intentionally disposed on level reset and **do not respawn** so the entity count now reports one fewer entity. Just wanted to mention this was by design.

Also, #185 is still open. We wanted to prioritise game-breaking bugs first but will tackle this in a follow-up PR!

jacobem1836 reviewed (approved):
Everything looks good now - self destruct drone works fine and bombs disposing correctly. Good stuff.

Russel1lTan committed:
Fixed SelfDestroyDrone's collider dispose and the damage of BombComponent. Solved BombComponentTest error

Russel1lTan committed:
Fixed SelfDestroyDrone's collider dispose and the damage of BombComponent. Solved BombComponentTest error

Russel1lTan committed:
Fixed SelfDestrcutComponent

charlottemaherr committed:
Restored changes overwritten in previous commit

charlottemaherr committed:
Add missing javadocs

charlottemaherr committed:
Restore version of lvl 1 game area same as main (plus bomber)

tristynhunt reviewed (approved):
changes look good and i believe all the bugs are fixed now :)
